enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/catchup/storecopy/FileSender.java

```diff
@@ -19,64 +19,89 @@
+    public boolean isEndOfInput() throws Exception
+    {
+        return endOfInput && preFetchedBytes == null && sentChunk;
+    }
+
+    @Override
+    public void close() throws Exception
+    {
+        channel.close();
+    }
+
+    @Override
+    public FileChunk readChunk( ByteBufAllocator allocator ) throws Exception
+    {
+        if ( isEndOfInput() )
+        {
+            return null;
+        }
+        else
+        {
+            sentChunk = true;
+        }
+
+        byte[] next = prefetch();
+        FileChunk fileChunk = FileChunk.create( preFetchedBytes == null ? new byte[0] : preFetchedBytes, next == null );
+
+        preFetchedBytes = next;
+
+        return fileChunk;
+    }
+
+    @Override
+    public FileChunk readChunk( ChannelHandlerContext ctx ) throws Exception
+    {
+        return readChunk( ctx.alloc() );
+    }
 
-    FileSender( FileSystemAbstraction fs, ChannelHandlerContext ctx )
+    @Override
+    public long length()
     {
-        this.fs = fs;
-        this.ctx = ctx;
+        return -1;
     }
 
-    void sendFile( File file ) throws IOException
+    @Override
+    public long progress()
     {
-        ctx.writeAndFlush( ResponseMessageType.FILE );
-        ctx.writeAndFlush( new FileHeader( file.getName() ) );
-        ByteBuffer buffer = ByteBuffer.allocateDirect( FileChunk.MAX_SIZE );
-        byte[] bytes = null;
-        try ( StoreChannel channel = fs.open( file, "r" ) )
+        return 0;
+    }
+
+    private byte[] prefetch() throws IOException
+    {
+        int bytesRead = channel.read( byteBuffer );
+        if ( bytesRead == -1 )
         {
-            boolean fileEmpty = true;
-            while ( channel.read( buffer ) != -1 )
-            {
-                fileEmpty = false;
-                // send out the previous chunk if there is any
-                perhapsWriteAndFlush( ctx, bytes, false );
-                // let's set the bytes to null in case we are not able to create another full array and overwrite it
-                bytes = null;
-
-                // if the buffer is full...
-                if ( !buffer.hasRemaining() )
-                {
-                    // create the next chunk
-                    bytes = createByteArray( buffer );
-                }
-            }
-
-            // if there are some more bytes left or the file is empty...
-            if ( buffer.position() > 0 || fileEmpty )
-            {
-                // send out the previous chunk if there is any
-                perhapsWriteAndFlush( ctx, bytes, false );
-                // and create a new one for the remaining bytes (or an empty byte array in case the file is empty)
-                bytes = createByteArray( buffer );
-            }
-
-            // send out whatever is left to send as the last chunk
-            perhapsWriteAndFlush( ctx, bytes, true );
+            endOfInput = true;
+            return null;
         }
+        return createByteArray( byteBuffer );
```

Comment:
Can we in future perhaps refer to the ByteBuf instead of byte[] in FileChunk? Not super important now.